docs/designs/ps_design.md

```diff
@@ -0,0 +1,429 @@
+    ... # keep the same with current code `GetModelRequest`
+}
+
+message PullModelResponse{
+    bool need_push_model = 1;
+    Model model = 2;
+}
+
+class PServer(elasticdl_pb2_grpc.PServerServicer):
+    ...
+    def pull_variable(self, request):
+        res = PullModelResponse()
+        if self._need_initialize_model:
+            res.need_push_model = True
+            return res
+        res.need_push_model = False
+        res.model = self._get_model() # get model in this PS node
+        return res
+
+def push_model(self, request):
+    model = request.model
+    ... # initialize model in this PS node
+		
+class Worker(object):
+    ...
+    def pull_variable(self):
+        # for-loop should be implemented in multithread
+        for ps_index in range(self._ps_node_num):
+            req = PullModelRequest() # create request code keeps the same with current code 
+            res = self._stub[ps_index].pull_variable() # pull variable from PS
+            if res.need_push_model:
+                // worker initializes its model here if needed
+                model = transform_worker's_model_to_protobuf()
+                self._stub[ps_index].push_model(model) # get model in this worker
+            req = PullModelRequest() # create request code keeps the same with current code
+            res = self._stub[ps_index].pull_variable() # pull variable from PS
+            if res.need_push_model:
+                raise Error or try a pre-defined constant times
+```
+
+Please Note that the worker only initializes model variables. ElasticDL adopts lazy initialization for embedding vectors. Please refer to "[Workers get model parameters from PS](#Workers-get-model-parameters-from-PS)" section.
+
+### Relaunch of PS
+In case a PS pod fails, the master will try to relaunch one PS and it should recover model variables and embedding tables.
+
+For model variables, PS can recover from workers in the same way as the variable initialization.
+
+For embedding tables, the `master.main` tells PS through in starting command that PS should recover from replica. If there is no replica, PS has to recover from checkpoint. More details can be find in section [Embedding Replicas in PS](#Embedding-Replicas-in-PS).
+
+### Workers get model variables from PS
+Before each forward-pass, workers need to get all model variables from PS. Currently, workers call function `get_model()` to get variables.
+
+Workers get embedding vectors from PS when the forward-pass function of the ElasticDL embedding layer is called. PS may not possess all the embedding vectors needed because ElasticDL adopts lazy initialization for embedding vectors, i.e. iniatializing embedding vectors when they are needed. Thus, if a worker wants to pull some embedding vectors that are not existing in PS, PS will create and initialize these embedding vectors and return their value to the worker.
+
+```python
+message EmbeddingTable{
+    string name;
+    Tensor embeddings;
+}
```

Comment:
There is a name field in Tensor already,  the name could be `embedding_layer0` or `embedding_layer0-momentum`. So, there is no need to add an extra EmbeddingTable message

Reply:
OK. I will modify this proto after discussion.